core/arch/arm/plat-sam/freq.c

```diff
@@ -12,8 +12,100 @@
 #include <kernel/panic.h>
 #include <libfdt.h>
 
+#define OPP_RATES_MAX 8
+
+/* the clock and its rates for CPU OPP */
+struct clk_rates {
+	struct clk *clk;
+	size_t rate_num; /* the number of valid clock rates in @rates */
+	unsigned long rates[OPP_RATES_MAX];
+};
+
+static struct clk_rates *opp_rates;
 static unsigned long freq;
 
+static TEE_Result get_rates_array(struct clk *clk, size_t start_index,
+				  unsigned long *rates, size_t *nb_elts)
+{
+	if (!opp_rates)
+		panic("Invalid CPU OPP Rates Array");
+
+	if (opp_rates->clk != clk)
+		panic("Invalid clock");
+
+	if (!rates) {
+		*nb_elts = opp_rates->rate_num;
+
+		return TEE_SUCCESS;
+	}
+
+	if (start_index > opp_rates->rate_num ||
+	    start_index + *nb_elts > opp_rates->rate_num) {
+		EMSG("Bad parameter(s): start_index %d, nb_elts %d",
+		     start_index, *nb_elts);
+
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	memcpy(rates, &opp_rates->rates[start_index],
+	       *nb_elts * sizeof(*rates));
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result dt_get_opp_hz(const void *fdt, int node, unsigned long *value)
+{
+	const char *property = "opp-hz";
+	const fdt64_t *p = NULL;
+	int len = 0;
+
+	p = fdt_getprop(fdt, node, property, &len);
+	if (!p)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (len != sizeof(*p))
+		return TEE_ERROR_BAD_FORMAT;
+
+	*value = fdt64_ld(p);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result opp_rates_setup(const void *fdt, int node, struct clk *clk)
+{
+	const char *compatible = "operating-points-v2";
+	const struct fdt_property *prop = NULL;
+	unsigned long rate = 0;
+	int opp_table = 0;
+	int offset = 0;
+
+	prop = fdt_get_property(fdt, node, compatible, NULL);
+	if (!prop)
+		return TEE_SUCCESS;
+
+	opp_rates = calloc(1, sizeof(*opp_rates));
+	if (!opp_rates) {
+		EMSG("Fail to alloc opp_rates");
+
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	offset = (int)prop - (int)fdt;
+	if (fdt_node_check_compatible(fdt, offset, compatible))
+		panic("Invalid opp-table");
+
+	fdt_for_each_subnode(opp_table, fdt, offset) {
+		if (dt_get_opp_hz(fdt, opp_table, &rate) != TEE_SUCCESS)
+			panic("Get opp-hz failed");
+
+		opp_rates->rates[opp_rates->rate_num++] = rate;
```

Comment:
prefer in 2 step an let compile optimize the whole (and check that index does not overflows).
Suggestion:
```c
	size_t rate_num = 0;

	(...)

	fdt_for_each_subnode(opp_table, fdt, offset) {
		if (rate_num >= OPP_RATES_MAX)
			panic("CPU OPPs overflow");

		if (dt_get_opp_hz(fdt, opp_table, opp_rates->rates + rate_num))
			panic("Get opp-hz failed");

		rate_num++;
	}
	opp_rates->rate_num = rate_num;

Reply:
Updated, thanks.